[AST] [Sema] [IDE] Fix places where we set bare arguments

beccadax opened the pull request:
This change adds an assertion that, except on `SelfApplyExpr`s, `Arg` must always be a `ParenExpr`, `TupleExpr`, `TupleShuffleExpr`, ~~`CodeCompletionExpr`, or `ErrorExpr`~~. It then fixes all failures caught by build-script -t or -T on macOS. I'm opening this PR now so I can let CI have a crack at it.

Along the way, I *may* have fixed rdar://41416911, but I'll need someone at Apple to confirm that for me.

I'm not sure if we should keep the assertions in the compiler or revert them and just merge the fixes, but I'm not at that stage yet anyway.

@CodaFi @slavapestov Probably relevant to your interests.

brentdax commented:
@swift-ci Please test

brentdax commented:
I'm a bit surprised by the lack of `CodeCompletionExpr`s and `ErrorExpr`s, but maybe it's true…

@swift-ci Please test

brentdax commented:
@swift-ci Please test

brentdax commented:
🤷‍♂️

@swift-ci Please test

CodaFi commented:
@rintaro also had some work on this in #17696

brentdax commented:
@CodaFi Yes, we seem to have implemented similar fixes. Looks like the difference is that I effectively `&&` the `CallArgs` test with the inner `if`, while @rintaro effectively `&&`s it with the outer `if`. I suspect he has it right and I have it wrong, but he might disagree.

If #17676 gets merged, I can back out that particular change, and I think the rest of this will slot right in.

slavapestov commented:
This looks like a great idea. Eventually we want ApplyExpr to directly store call arguments and encode which ones are packed into variadics and defaulted, etc. Normalizing the AST in this manner is a good first step and also makes sense on its own.

brentdax commented:
@swift-ci please test

brentdax commented:
🙄

@swift-ci please test

beccadax committed:
[AST] Don’t allow application of most bare exprs

Most ApplyExpr subclasses will now require a ParenExpr, TupleExpr, or TupleShuffleExpr around their argument lists. The exceptions are BinaryExpr, which requires a TupleExpr, and SelfApplyExpr and its subclasses, which allow anything (and only ever have one argument).

This change doesn’t fix any of the places where we actually generate these.

beccadax committed:
[Sema] Respect other ApplyExprs’ call parens

PreCheckExpression was not treating the implicit call parens on unary or binary operator applications like the explicit parens on CallExpr applications. Now it will. This prevents various forms of mischief, like having them subsumed into a TypeExpr.

beccadax committed:
[Sema] Add arg parens to two generated calls

beccadax committed:
[IDE] Add call parens to postfix operator completion

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

rintaro commented on include/swift/AST/Expr.h:
```diff
@@ -5128,6 +5133,15 @@ inline bool Expr::isInfixOperator() const {
+  else if (isa<BinaryExpr>(this))
+    return isa<TupleExpr>(e);
+  else
+    return isa<ParenExpr>(e) || isa<TupleExpr>(e) || isa<TupleShuffleExpr>(e);
```
Is it worth to add `PrefixUnaryExpr`/`PostfixUnaryExpr` -> `ParenExpr` check here? cc: @xedin

xedin replied:
Since arguments of such expressions are _now_ implicitly wrapped into parens it shouldn't be required.

rintaro replied:
That's why I thought we *can* add the check for it.

brentdax replied:
I don’t think we should. To be honest, I’d like to drop the BinaryExpr -> TupleExpr check too, because I want to support operator functions with extra defaulted parameters (for #line and friends). But that will require other changes to the compiler, and I’d rather do one thing at a time.

brentdax commented:
@swift-ci please smoke test and merge
